File: src/Server_multi.py
#
# Written by kuwingto, 10 March 2020
#

import os
import math
import random
import numpy as np
import cv2
import importlib.util

import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import sys
import time

import zmq
import my_message_multi_pb2

context = zmq.Context()
socket = context.socket(zmq.REP)
socket.bind("tcp://*:5555")


class segmentationServer:
    def __init__(self):
        self.setupNetwork()
        print("Ready")
    def setupNetwork(self):
        # setup your network here. 
        self.a = 1
        
        
if __name__ == '__main__' :
    from time import time
    ser = segmentationServer()
    
     # Here For demonstration, we read a image to the buffer and will send back to CLient
    outimage = cv2.imread("../data/segmentation.bmp", cv2.IMREAD_GRAYSCALE)
    
    while True:
        print("waiting connection")
        message = socket.recv()
        print("message recv")
        if (message):
            start = time()
            
            #input images
            my_message_input = my_message_multi_pb2.my_message_multi()
            my_message_input.ParseFromString(message)
            for i in range(len(my_message_input.imgs) ):
                print("received image width = ", my_message_input.imgs[i].width)
                print("received image height = ", my_message_input.imgs[i].height)
                #print(type(my_message_input.imgs[i].image_data))
                nparr = np.frombuffer(my_message_input.imgs[i].image_data,dtype=np.ubyte)
                print("data shape = ",nparr.shape)
                img = nparr.reshape(my_message_input.imgs[i].height,my_message_input.imgs[i].width)
                print("image shape = ",img.shape)
                
                #uncomment show the image
                #print(img.shape)
                #stacked_img = np.stack((img,)*3, axis=-1)
                #imgplot = plt.imshow(stacked_img)
                #plt.show()
            
            #handling 
            #???
                
            #output
            my_message_output = my_message_multi_pb2.my_message_multi()
            class_id = my_message_input.class_id
            height = outimage.shape[0]
            width = outimage.shape[1]
            sendimage = outimage
            
            #prepare message
            my_message_output.class_id = my_message_input.class_id
            
            # 1st image
            img = my_message_output.imgs.add()
            img.width = width
            img.height = height
            img.image_data = sendimage.tobytes() 

             # 2nd image
            img = my_message_output.imgs.add()
            img.width = width
            img.height = height
            img.image_data = sendimage.tobytes() 
            
            #serialize message
            str = my_message_output.SerializeToString()
            socket.send(str)
        else:
            print("error message")

File: src/server_multi.cpp
/* 
* Written by kuwingto, 10 March 2019
*/

#include <string.h>
#include <stdio.h>
#include <chrono>  // for high_resolution_clock
#include <math.h>
#include <opencv2/opencv.hpp>
#include <zmq.hpp>
#include "my_message_multi.pb.h"

int main()
{
	GOOGLE_PROTOBUF_VERIFY_VERSION;
	zmq::context_t m_context = zmq::context_t(1);
	zmq::socket_t *m_pSock;

	/****** start service **********/
	m_pSock = new zmq::socket_t(m_context, ZMQ_REP); // rep for server
	m_pSock->bind("tcp://*:5555");

	/************ service loop ***************/
	while (true)
	{
		std::cout << "waiting connection...\n";
		zmq::message_t request;
		if (m_pSock->recv(&request, 0))
		{
			std::cout << "receive message...\n";
			/************ receive message ***************/
			std::string  msgStr = std::string((char*)request.data(), request.size());
			//unserialize
			my_message_multi out;
			out.ParseFromString(msgStr);

			int class_id = out.class_id();
			std::cout << "class_id = " << class_id << "\n";
			for (int i = 0; i < out.imgs_size(); ++i)
			{
				int width = out.imgs(i).width();
				int height = out.imgs(i).height();
				std::cout << "width = " << width << " height = " << height << "\n";
				cv::Mat img = cv::Mat(height, width, CV_8UC1);
				memcpy(img.data, &out.imgs(i).image_data()[0], sizeof(uchar) * width * height);

				cv::imshow("img" + std::to_string(i), img);
				cv::waitKey(1);
			}

			/************ prepare reply ***************/
			cv::Mat img;
			img = cv::imread("../data/test.bmp", CV_LOAD_IMAGE_GRAYSCALE);

			my_message_multi sendPack;
			sendPack.set_class_id(1);

			my_message_multi_Mat* sendMat = sendPack.add_imgs();
			(*sendMat).set_width(img.size().width);
			(*sendMat).set_height(img.size().height);
			(*sendMat).set_image_data((char *)img.data, sizeof(uchar) * img.size().width * img.size().height);
			std::string s1 = (*sendMat).SerializeAsString();
			std::cout << "s1.size() = " << s1.size() << "\n";

			std::string s = sendPack.SerializeAsString();
			std::cout << "s.size() = " << s.size() << "\n";
			/************ send reply  ***************/
			{
				zmq::message_t message(s.size());
				memcpy(message.data(), s.data(), s.size());
				m_pSock->send(message);
			}
		}
	}

	std::cout << "press to continue \n";
	getchar();

	m_pSock->close();
	delete m_pSock;
    return 0;
}

